src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs

```diff
@@ -0,0 +1,212 @@
+        }
+
+        private static bool IsTestMethod(MethodSymbol method)
+        {
+            // ToDo PROTOTYPE (https://github.com/dotnet/roslyn/issues/9811): Make this real. 
+            return method.Name.StartsWith("Test");
+        }
+    }
+
+    internal sealed class InstrumentationInjectionRewriter : BoundTreeRewriterWithStackGuard
+    {
+        private readonly MethodSymbol _method;
+        private readonly ArrayBuilder<SourceSpan> _spansBuilder;
+        private readonly FieldSymbol _payload;
+        private readonly TypeCompilationState _compilationState;
+        private readonly DiagnosticBag _diagnostics;
+        private readonly DebugDocumentProvider _debugDocumentProvider;
+
+        public static void InstrumentMethod(MethodSymbol method, BoundBlock methodBody, FieldSymbol payloadField, TypeCompilationState compilationState, DiagnosticBag diagnostics, DebugDocumentProvider debugDocumentProvider, out ImmutableArray<SourceSpan> dynamicAnalysisSpans, out BoundBlock newMethodBody)
+        {
+            ArrayBuilder<SourceSpan> spansBuilder = ArrayBuilder<SourceSpan>.GetInstance();
+            BoundTreeRewriter collector = new InstrumentationInjectionRewriter(method, spansBuilder, payloadField, compilationState, diagnostics, debugDocumentProvider);
+            newMethodBody = (BoundBlock)collector.Visit(methodBody);
+            dynamicAnalysisSpans = spansBuilder.ToImmutableAndFree();
+        }
+
+        private InstrumentationInjectionRewriter(MethodSymbol method, ArrayBuilder<SourceSpan> spansBuilder, FieldSymbol payload, TypeCompilationState compilationState, DiagnosticBag diagnostics, DebugDocumentProvider debugDocumentProvider)
+        {
+            _method = method;
+            _spansBuilder = spansBuilder;
+            _payload = payload;
+            _compilationState = compilationState;
+            _diagnostics = diagnostics;
+            _debugDocumentProvider = debugDocumentProvider;
+        }
+
+        public override BoundNode Visit(BoundNode node)
+        {
+            BoundStatement statement = node as BoundStatement;
+            if (statement != null)
+            {
+                switch (statement.Kind)
+                {
+                    case BoundKind.SwitchSection:
+                    case BoundKind.SwitchLabel:
+                    case BoundKind.PatternSwitchSection:
+                    case BoundKind.PatternSwitchLabel:
+                    case BoundKind.CatchBlock:
+                    // A labeled statement or a sequence point can be ignored with respect to instrumentation in favor of the underlying statement.
+                    case BoundKind.SequencePoint:
+                    case BoundKind.SequencePointWithSpan:
+                    case BoundKind.LabeledStatement:
+                        break;
+                    default:
+                        return CollectDynamicAnalysis(base.Visit(node));
+                }
```

Comment:
Also, for my education, what was the reason to do this `switch` rather than directly do the check that appears below?

```CS
      BoundStatement statement = node as BoundStatement;
           if (statement != null)
```